### PR TITLE
[jvm-packages] add batchSize to xgboost4j and xgboost-spark

### PR DESCRIPTION
**Pull request by chenqin**
per feedback from [#3701](https://github.com/dmlc/xgboost/pull/3701)

### CODE REVIEW

**Comment by chenqin** [comment]
@CodingCat ping

**Comment by CodingCat** [comment]
@yanboliang can you do me a favor to look at this? I am currently working on something else

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 1]
```diff
@@ -378,11 +378,13 @@ private object Watches {
     }
 
     val (trainIter1, trainIter2) = trainPoints.duplicate
-    val trainMatrix = new DMatrix(trainIter1, cacheDirName.map(_ + "/train").orNull)
+    val trainMatrix = new DMatrix(trainIter1, cacheDirName.map(_ + "/train").orNull,
+      params.get("batchSize").map(_.toString.toInt).getOrElse(32<<10))
```
Comment (yanboliang) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: If we set default value at ```BoosterParams```, here we can simplify as: ```params.get("batchSize").toInt```. We should manage default value at a fixed place to make the maintenance more easy.

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifierSuite.scala [hunk 2]
```diff
@@ -35,7 +35,8 @@ class XGBoostClassifierSuite extends FunSuite with PerTest {
       "eta" -> "1",
       "max_depth" -> "6",
       "silent" -> "1",
-      "objective" -> "binary:logistic")
+      "objective" -> "binary:logistic",
+      "batchSize" -> "200")
```
Comment (yanboliang) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifierSuite.scala]: In the key-value way to set params, it should be ```batch_size```.

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/DMatrix.java [hunk 3]
```diff
@@ -36,18 +36,17 @@
...
+     * @param cacheInfo Cache path information, used for external memory setting, can be null.
+     * @param batchSize Size of batch, recommended 32 << 10
+     * @throws XGBoostError
+     */
+  public DMatrix(Iterator<LabeledPoint> iter, String cacheInfo, int batchSize) throws XGBoostError {
     if (iter == null) {
       throw new NullPointerException("iter: null");
     }
-    // 32k as batch size
-    int batchSize = 32 << 10;
```
Comment (yanboliang) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/DMatrix.java]: This means if users call XGBoost Java API to train a model, then there is no default batch size; but there is default value for Scala API. Is this by design? I think here we should follow the Scala way to set default value in arguments.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/BoosterParams.scala [hunk 4]
```diff
@@ -99,6 +99,15 @@ private[spark] trait BoosterParams extends Params {
...
   final def getSubsample: Double = $(subsample)
 
+  /**
+    * batch size of training instance. Setting 32 << 10 means XGBoost collect 32k rows at once
+    * and construct DMatrix in batch fashion.
+    */
+  final val batchsize = new IntParam(this, "batchSize", "DMatrix batch size of the training " +
+    "insance. Setting it to 32 << 10 means that building a DMatrix in micro batch",
+    (value: Int) => value > 0
+  )
```
[the remark this answers is not in the document]
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/BoosterParams.scala]: done

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifierSuite.scala (hunk 2 above)
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifierSuite.scala]: done

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/DMatrix.java (hunk 3 above)
Comment (chenqin) [reply jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/DMatrix.java]: added constructor with default batch size

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 1 above)
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: sounds good, fixed.

**Comment by chenqin** [comment]
per feedback make some changes.

**Comment by chenqin** [comment]
There seems a issue in CheckpointManagerSuite where core don't have batchsize as parameter saved in model.  I am hesitate to make change in core and jvm at same time. Hence I would like to add TODO on [toInt](https://github.com/dmlc/xgboost/pull/3705/files#diff-095250777d030169ae83818b37ab44b1R386) and fall back to set value implementation.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 5]
```diff
@@ -378,11 +378,13 @@ private object Watches {
...
 
     val (trainIter1, trainIter2) = trainPoints.duplicate
-    val trainMatrix = new DMatrix(trainIter1, cacheDirName.map(_ + "/train").orNull)
+    val trainMatrix = new DMatrix(trainIter1, cacheDirName.map(_ + "/train").orNull,
+      params.get("batchSize").toString.toInt)
     val trainGroups = buildGroups(trainIter2.map(_.group).toSeq).toArray
     trainMatrix.setGroup(trainGroups)
 
-    val testMatrix = new DMatrix(testPoints.iterator, cacheDirName.map(_ + "/test").orNull)
+    val testMatrix = new DMatrix(testPoints.iterator, cacheDirName.map(_ + "/test").orNull,
```
Comment (chenqin) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: would like to add todo and make it backward compatible with saved model without batch_size parameter from core. TODO for next pull, make change to core and include batch_size in model serialization

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/BoosterParams.scala [hunk 6]
```diff
@@ -244,7 +255,7 @@ private[spark] trait BoosterParams extends Params {
 
   setDefault(eta -> 0.3, gamma -> 0, maxDepth -> 6,
     minChildWeight -> 1, maxDeltaStep -> 0,
-    growPolicy -> "depthwise", maxBins -> 16,
+    growPolicy -> "depthwise", maxBins -> 16, batchsize -> (32 << 10),
```
Comment (yanboliang) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/BoosterParams.scala]: ```batchsize``` -> ```batchSize```. There are some historical reasons, but new added params should follow correct convention.

**Comment by chenqin** [comment]
CI seems flaky and irrelevant.
continuous-integration/travis-ci/pr caused by
`mkl-random 1.0.1 requires cython, which is not installed.
mkl-fft 1.0.4 requires cython, which is not installed.`

appVeyor error 
`
Build started
git clone -q https://github.com/dmlc/xgboost.git C:\projects\xgboost
fatal: unable to access 'https://github.com/dmlc/xgboost.git/': Failed to connect to github.com port 443: Timed out
Command exited with code 128
if /i "%target%" == "rmingw" ( 7z a failure.zip *.Rcheck\* && appveyor PushArtifact failure.zip )
7-Zip [64] 16.04 : Copyright (c) 1999-2016 Igor Pavlov : 2016-10-04
`

**Comment by hcho3** [comment]
@chenqin I restarted the failed tests for now. In the near future, I'm going to look at disabling flaky tests entirely.

**Comment by hcho3** [comment]
@chenqin Please ignored the failed test `continuous-integration/jenkins/pr-head`. (This was due to misconfiguration of Jenkins CI server.) Only `continuous-integration/jenkins/pr-merge` is relevant.

**Comment by chenqin** [comment]
Got it, do we have another pr track misconfiguration fix? Also, seems tests passed.
> @chenqin Please ignore the failed test `continuous-integration/jenkins/pr-head`. (This was due to misconfiguration of Jenkins CI server.) Only `continuous-integration/jenkins/pr-merge` is relevant.

**Comment by hcho3** [comment]
@chenqin I don't know much about Java. Let's see if @yanboliang and @CodingCat are okay with merging this.

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifierSuite.scala [hunk 7]
```diff
@@ -94,6 +95,7 @@ class XGBoostClassifierSuite extends FunSuite with PerTest {
     val model2 = new XGBoostClassifier()
       .setEta(1)
       .setMaxDepth(6)
+      .setBatchSize(200)
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifierSuite.scala]: why add batchSize in this test?

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifierSuite.scala [hunk 8]
```diff
@@ -35,7 +35,8 @@ class XGBoostClassifierSuite extends FunSuite with PerTest {
       "eta" -> "1",
       "max_depth" -> "6",
       "silent" -> "1",
-      "objective" -> "binary:logistic")
+      "objective" -> "binary:logistic",
+      "batch_size" -> "200")
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifierSuite.scala]: also here, it should be in the test syncing parameters?

File: jvm-packages/xgboost4j/src/test/java/ml/dmlc/xgboost4j/java/DMatrixTest.java [hunk 9]
```diff
@@ -42,12 +42,20 @@ public void testCreateFromDataIterator() throws XGBoostError {
       blist.add(p);
       labelall.add(p.label());
     }
-    DMatrix dmat = new DMatrix(blist.iterator(), null);
+
+    DMatrix dmat = new DMatrix(blist.iterator(), null, 32<<10);
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/test/java/ml/dmlc/xgboost4j/java/DMatrixTest.java]: unnecessary change as the default value is 32<<10?

**Comment by CodingCat** [comment]
close for now
